### PR TITLE
ZStream drop and take improvements

### PR DESCRIPTION
**Pull request by vasilmkd**
@iravid 

@adamgfraser I'd like to ask you specifically about the changes I had to make about the widening warnings. Are they sufficient or would you like to go further (accepting `Long` type arguments instead of `Int`).

### CODE REVIEW

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 1]
```diff
@@ -2270,15 +2270,15 @@ class ZStream[-R, +E, +A] private[stream] (private[stream] val structure: ZStrea
...
+  def take(n: Long): ZStream[R, E, A] =
     ZStream {
       for {
         as      <- self.process
-        counter <- Ref.make(0).toManaged_
-        pull = counter.modify { c =>
-          if (c >= n) (Pull.end, c)
-          else (as, c + 1)
-        }.flatten
+        counter <- Ref.make(0L).toManaged_
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala]: Was it necessary to change the formulation here?

**Comment by vasilmkd** [comment]
@jdegoes Could you please review this PR? 🙏

File: streams/shared/src/main/scala/zio/stream/ZStream.scala (hunk 1 above)
Comment (regiskuckaertz) [reply streams/shared/src/main/scala/zio/stream/ZStream.scala]: I think it does if we adopt the stance that failed elements are not counted in, but there needs to be a fallback for when an element fails.
```scala
for {
  as <- self.process
  counter <- Ref.make(n).toManaged_
  pull = ref.get {
    case 0 => Pull.end
    case n => as.foldM({
      case None => Pull.end
      case _ => pull
    },
    a => counter.update(_ - 1) *> Pull.emit(a)
  }
yield pull
```

Comment (vasilmkd) [reply streams/shared/src/main/scala/zio/stream/ZStream.scala]: This here swallows the error right? I think we do need to propagate it downstream, the user will have enough power to deal with it in the future.

Comment (regiskuckaertz) [reply streams/shared/src/main/scala/zio/stream/ZStream.scala]: or in the past 🤗 with `stream.either.take(n).absolve`. I also think it's more intuitive that "take the first n" implies the first n _successful_ elements in this context, as errors are only first-class when they are inside the value channel.

Comment (vasilmkd) [reply streams/shared/src/main/scala/zio/stream/ZStream.scala]: So how does this translate to the "counts (or doesn't) towards the `n` budget" debate?

Comment (regiskuckaertz) [reply streams/shared/src/main/scala/zio/stream/ZStream.scala]: The version above doesn't count errors in the budget.

Comment (mschuwalow) [reply streams/shared/src/main/scala/zio/stream/ZStream.scala]: That version would also run into issues if the user provided n is negative.
I agree with the only considering successful elements though.
Excited to see where this is going

**Comment by regiskuckaertz** [comment]
Looks like you need to annotation the type of the stream

**Comment by vasilmkd** [comment]
Happy New Year! @regiskuckaertz

**Pull request merged**